### PR TITLE
Retry enhancements

### PR DESCRIPTION
**Pull request by rossabaker**

### CODE REVIEW

File: client/src/main/scala/org/http4s/client/middleware/Retry.scala [hunk 1]
```diff
@@ -2,71 +2,141 @@ package org.http4s
...
+    else false
+  }
+
+  /** Returns true if the request method is idempotent and the result is
+    * either a throwable or has one of the `RetriableStatuses`.  This is
+    * the `defaultRetriable` behavior from 0.16.
+    * 
+    * Caution: if the request body is effectful, the effects will be
+    * run twice.  The most common symptom of this will be resubmitting
+    * an empty request body.
```
Comment (rossabaker) [inline client/src/main/scala/org/http4s/client/middleware/Retry.scala]: This is a really horrid thing to do, but some people will want this.  An alternative is to buffer request bodies, which would slow things down.

File: client/src/main/scala/org/http4s/client/middleware/Retry.scala [hunk 2]
```diff
@@ -2,71 +2,141 @@ package org.http4s
...
+  /** Like [[unsafeRetriable]], but returns true even if the request method
+    * is not idempotent.  This is useful if failed requests are assumed to
+    * have not reached their destination, which is a dangerous assumption.
+    * Use at your own risk.
+    * 
+    * Caution: if the request body is effectful, the effects will be
+    * run twice.  The most common symptom of this will be resubmitting
+    * an empty request body.
+    */
+  def recklesslyRetriable(req: Request, result: Either[Throwable, Response]): Boolean =
```
Comment (ChristopherDavenport) [inline client/src/main/scala/org/http4s/client/middleware/Retry.scala]: Best Ever!

File: client/src/test/scala/org/http4s/client/middleware/RetrySpec.scala [hunk 3]
```diff
@@ -21,24 +21,25 @@ class RetrySpec extends Http4sSpec with Tables {
   }
 
   val defaultClient = Client.fromHttpService(route)
+  val failClient = Client(Service.const(Task.fail(new Exception("boom"))), Task.now(()))
```
Comment (ChristopherDavenport) [inline client/src/test/scala/org/http4s/client/middleware/RetrySpec.scala]: Are we using the failClient anywhere that required it's removal?

File: client/src/main/scala/org/http4s/client/middleware/Retry.scala [hunk 4]
```diff
@@ -2,71 +2,141 @@ package org.http4s
...
       }
     }
 
-    def nextAttempt(req: Request, attempts: Int, duration: FiniteDuration): Task[DisposableResponse] = {
-      // TODO honor Retry-After header
-      prepareLoop(req.withEmptyBody, attempts + 1).schedule(duration)
+    def nextAttempt(req: Request, attempts: Int, duration: FiniteDuration, retryHeader: Option[`Retry-After`]): Task[DisposableResponse] = {
+      val headerDuration = retryHeader.map { h =>
+        h.retry match {
+          case Left(d) => Instant.now().until(d.toInstant, ChronoUnit.SECONDS)
```
Comment (ChristopherDavenport) [inline client/src/main/scala/org/http4s/client/middleware/Retry.scala]: This is the decreasing time location?

File: client/src/test/scala/org/http4s/client/middleware/RetrySpec.scala (hunk 3 above)
Comment (rossabaker) [reply client/src/test/scala/org/http4s/client/middleware/RetrySpec.scala]: I think I started another block for more tests and moved it and then abandoned that block.

File: client/src/main/scala/org/http4s/client/middleware/Retry.scala (hunk 4 above)
Comment (rossabaker) [reply client/src/main/scala/org/http4s/client/middleware/Retry.scala]: Yeah, if a Retry-After is a Left, we need to calculate the delta from now until then.  The Right is expressed as a wait time in seconds.

**Pull request merged**